### ISSUE TITLE
[Feature] watch mode for e2e tests

### ISSUE THREAD

**Issue by JoelEinbinder** [microsoft/playwright#7035]
I'd really appreciate a watch mode. My hands hurt from running my test over and over.

---

If you are looking for the watch mode that runs tests on save, I found assigning the shortcut to Run Test at Cursor very handy:
<img width="562" alt="Screenshot 2022-12-01 at 7 44 45 AM" src="https://user-images.githubusercontent.com/883973/205097424-10a0b9bd-148b-4379-bb58-693d5eba6e11.png">

## [Playwright VS Code Extension](https://marketplace.visualstudio.com/items?itemName=ms-playwright.playwright) is likely to mitigate your request 

VS Code extension makes writing and debugging tests convenient, there is no need to go between code and terminal back and forth.

**Comment by pavelfeldman** [comment]
This is going to be rebooted as a part of the debugging story. Closing with no replacement for now.

**Comment by iamyoki** [comment]
> As a workaround _with Playwright 1.12+_ I’m using the `test:e2e:watch` npm script below to watch all my test files and run a single test file when it is modified (install [chokidar-cli](https://github.com/kimmobrunfeldt/chokidar-cli) before):
> 
> ```json
> { "scripts": {
>     "test:e2e": "playwright test --config=tests/playwright.config.ts",
>     "test:e2e:watch": "chokidar 'tests/e2e/**/*.test.ts' -c 'npm run test:e2e -- --retries=0 {path}'",
> }}
> ```
> 
> Of course modify these scripts according to your needs. Coupled to `test.only` it’s perfect for my usage.

This would be better :)
`"test:e2e:watch": "chokidar e2e/**/*.{spec,test}.{js,ts} -c 'yarn run test:e2e'"`

**Comment by iamyoki** [comment]
I made a library `playwright-watch` to achieve this.
See [playwright-watch docs](https://github.com/iamyoki/playwright-watch)

<h1 align="center">playwright-watch</h1>

<div align="center">
  <a href="https://www.npmjs.com/package/playwright-watch">
    <img src="https://img.shields.io/npm/v/playwright-watch" alt="npm">
  </a>

  <a href="https://www.npmjs.com/package/playwright-watch">
    <img src="https://img.shields.io/npm/v/playwright-watch/latest" alt="npm">
  </a>

  <a href="https://github.com/iamyoki/playwright-watch">
    <img src="https://img.shields.io/npm/l/playwright-watch" alt="github">
  </a>

  <a href="https://github.com/iamyoki/playwright-watch">
    <img src="https://img.shields.io/github/issues/iamyoki/playwright-watch" alt="github issues">
  </a>
</div>

<br>
<br>

<p align="center">
  <code>playwright</code>
  <code>+</code>
  <code>watch</code>
  <code>=</code>
  <code>palywright-watch</code>
</p>
<p align="center">🙈🐵 Run Playwright in watch mode</p>

<br>

## Usage

use exactly the same as Playwright, only in watch mode
```bash
npx playwright-watch test
```

> Note: There is a peer dependency on `playwright`.

### Or you can install locally

```bash
yarn add playwright-watch -D

#or

npm install playwright-watch --save-dev
```

Add script to your package.json

```json
{
  "test:watch": "playwright-watch test"
}
```

or

```json
{
  "test": "playwright test",
  "test:watch": "playwright-watch exec yarn test"
}
```

## CLI Details

run `npx playwright-watch --help` to see more details.

```bash
🙈 playwright-watch

🐵 Run Playwright in watch mode

Commands:
  playwright-watch                    Use the same as Playwright but under
                                      watch
                                      (e.g. playwright-watch test)   [default]
  playwright-watch exec <commands..>  Run commands when changes are detected
                                      (e.g. playwright-watch exec yarn test)

Options:
  -h, --help        Show help                                        [boolean]
  -v, --version     Show version number                              [boolean]
  -c, -c, --config  Configuration file of Playwright (default:
                    playwright.config.js), the `testMatch` prop will be watch
                    path (default: `.*(test|spec).(js|ts|mjs)`).

Examples:
  playwright-watch test
  playwright-watch test --config tests/playwright.config.js

  playwright-watch exec yarn playwright test
  playwright-watch exec yarn run test:e2e
  playwright-watch exec echo changed
```

## Author

[Yoki Yu (iamyoki)](https://github.com/iamyoki)

**Comment by schickling** [comment]
Thanks for this tool @iamyoki as a workaround!

@pavelfeldman would be great if adding a watch mode could still be considered to be added to Playwright regardless. I’m planning on using PW as a general test runner (mostly for non-web tests), so this functionality would be quite important for my use cases. 🙏

**Comment by YonatanKra** [comment]
@pavelfeldman 
We really need a watch mode that also leaves the browsers open.
I'd be happy to work on this if you could guide me.

**Comment by frederikhors** [comment]
> @pavelfeldman We really need a watch mode that also leaves the browsers open. I'd be happy to work on this if you could guide me.

Yes please!

**Comment by dannyyassine-ce** [comment]
until then, another workaround using `nodemon`: 
```shell
nodemon --watch playwright/specs --exec \"playwright test -c playwright.config.ts\"",
```

also works great when using with the inspector:
```shell
nodemon --watch playwright/specs --exec \"PWDEBUG=1 playwright test -c playwright.config.ts\"",
```

**Comment by dannyyassine-ce** [comment]
> > until then, another workaround using `nodemon`:
> > ```shell
> > nodemon --watch playwright/specs --exec \"playwright test -c playwright.config.ts\"",
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > also works great when using with the inspector:
> > ```shell
> > nodemon --watch playwright/specs --exec \"PWDEBUG=1 playwright test -c playwright.config.ts\"",
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > if using with typescript with custom tsconfig.json, I have set up a `pw:dev` script in my package.json to watch when `tsc` changes the output directory to restart `playwright`:
> > ```json
> > "pw:dev": "yarn pw:watch & nodemon --watch playwright/tests-out --exec \"yarn pw:test\"",
> > "pw:dev:debug": "yarn pw:watch & nodemon --watch playwright/tests-out --exec \"yarn pw:test:debug\"",
> > "pw:watch": "tsc --incremental -p playwright/specs/tsconfig.json --watch & tsc-alias -p playwright/specs/tsconfig.json -w",
> > "pw:test": "playwright test -c playwright.config.ts",
> > "pw:test:debug": "PWDEBUG=1 playwright test -c playwright.config.ts"
> > ```
> 
> Before I decided to create [playwright-watch](https://github.com/iamyoki/playwright-watch), I have tried this way, the only problem is that if the save is triggered multiple times in a short period of time, playwright will concurrently run multiple processes and cause memory overflow.

you can always add a delay with changing files :) -> https://github.com/remy/nodemon#delaying-restarting

**Comment by cliffordfajardo** [comment]
We recently started using Playwright where I work & many peer teams are starting to move away from Cypress
I don't have much Cypress experience, I started off using Playwright

1 common thing I hear from former Cypress users is if there is something similar to the `npm run cy:open` command which opens the browser (non-headless) and watches for file changes
Simply adding as a data point ^

Random Aside:
If you have feel like Playwright is missing some features which you think are relavant to Cypress, please leave a comment here: https://github.com/microsoft/playwright/discussions/11201

**Comment by iamzapata** [comment]
Hey, adding a bit to the discussion here. About 3 years ago I build a large e2e test suite with Cypress and I was in love with it. After a long time battling with cypress I came to also hate some parts of it and started looking for alternatives and found playwright. 

This seems like a winner, but I just can't use it. It's impossible to be productive when writing new specs if I have to constantly type commands go back and forth between the test code and the test runner. 

I haven't read all of the documentation yet, but is there something I'm missing? I think watch mode should be one of the things that should just work out of the box.

**Comment by timcash** [comment]
> @pavelfeldman We really need a watch mode that also leaves the browsers open. I'd be happy to work on this if you could guide me.

This is needed, tests should be able to
1.) all run in the same instance of an open browser
2.) reload on test change
3.) reload on dependency change

**Comment by pavelfeldman** [comment]
@iamzapata I think you want this: https://github.com/microsoft/playwright-vscode

**Comment by franzvezuli** [comment]
A watch mode would be great, it's kind of the only deal-breaker between Playwright and Cypress in my opinion.

**Comment by bichotll** [comment]
It would be nice if meanwhile the `nodemon` solution or similar were added to the documentation :)

**Comment by Kerren-Entrostat** [comment]
Hi everyone, I just wanted to add some stuff here for `nodemon` if anyone is looking for a quick refresh that only runs the test you change.

[@iamyoki](https://github.com/iamyoki) I think your CLI is really cool and I'll probably use it, but just for if someone cannot install more deps or wants to use `shell` to achieve this, this is the `package.json` I have:

```json
{
    "scripts": {
         "test:locally:watch": "nodemon --verbose --signal SIGHUP --ext ts --exec \"npm run _test:locally:watch\"",
        "_test:locally:watch": "echo \"Running tests: \n$(find tests -newermt '5 seconds ago')\" && playwright test $(find tests -newermt '5 seconds ago' | tr '\\n' ' ')"
    }
}
```

Basically, all this does is checks which files changed in the last 5 seconds and runs them, it's not the best solution but it should work in most cases.

**Comment by einarpersson** [comment]
Please add watch mode to the CLI. I cannot see any reason against it, other than forcing users to vscode. Although I myself use vscode, I don't want to force it upon others. Please Microsoft, be kind and generous.

I want to use playwright for both my e2e tests and unit tests (and for the latter watch mode is a must).

**Comment by andrewmclagan** [comment]
oh wow there is actually no watch mode... huh.

**Comment by alexbjorlig** [comment]
Correct me if I'm wrong, but if we use `vitest` like shown [here](https://github.com/vitest-dev/vitest/blob/09d6222669433423a916d086022972761415d245/examples/playwright/test/basic.test.ts#L26) - how do we then benefit from playwright fixtures extending the base for example?

**Comment by zoltan-nz** [comment]
@alexbjorlig You totally right, if you use different test runners, you get the third party test runner's benefits, but loose the playwright test runner underlying features. In terms of fixtures, we have to directly setup page fixtures, maybe in the `beforeAll`. Hopefully watch mode will be implemented in Playwright eventually. Playwright is such an amazing tool. Love it.

**Comment by SalahAdDin** [comment]
It is still in the plans, right?

**Comment by mdee77** [comment]
Watch mode is a must have feature in my opinion. Have been excited about Playwright and had plans to switch from Cypress, but not without having a watch mode.

**Comment by SalahAdDin** [comment]
> Watch mode is a must have feature in my opinion. Have been excited about Playwright and had plans to switch from Cypress, but not without having a watch mode.

Well, It is late for us, we switched to Playwright not knowing there was not a watch mode.

**Comment by sand4rt** [comment]
Be patient guys, i think they will add it sooner or later!

**Comment by MikeShi42** [comment]
Hey everyone! We’ve hacked together a visual watch mode for Playwright - it’ll run and pop open a debugger any time you hit save (similar to Cypress) - would love to hear what y’all think!

[https://www.npmjs.com/package/@deploysentinel/playwright-watch](https://www.npmjs.com/package/@deploysentinel/playwright-watch)

![pw-demo-lg](https://user-images.githubusercontent.com/2781687/218813026-e8c6ee18-2f03-4228-ba55-d8fa4b852065.gif)

### tl;dr docs: 
Install:
```
npm i -D @deploysentinel/playwright-watch @deploysentinel/playwright
``` 
Run:
```
npx pw-watch --reporter @deploysentinel/playwright
```

**Comment by nil1511** [comment]
I guess [watch mode](https://github.com/microsoft/playwright/pull/20647) is coming soon

**Comment by SalahAdDin** [comment]
> Hey everyone! We’ve hacked together a visual watch mode for Playwright - it’ll run and pop open a debugger any time you hit save (similar to Cypress) - would love to hear what y’all think!
> 
> https://www.npmjs.com/package/@deploysentinel/playwright-watch
> 
> ![pw-demo-lg](https://user-images.githubusercontent.com/2781687/218813026-e8c6ee18-2f03-4228-ba55-d8fa4b852065.gif) [ ![pw-demo-lg](https://user-images.githubusercontent.com/2781687/218813026-e8c6ee18-2f03-4228-ba55-d8fa4b852065.gif) ](https://user-images.githubusercontent.com/2781687/218813026-e8c6ee18-2f03-4228-ba55-d8fa4b852065.gif) [ ](https://user-images.githubusercontent.com/2781687/218813026-e8c6ee18-2f03-4228-ba55-d8fa4b852065.gif)
> ### tl;dr docs:
> 
> Install:
> 
> ```
> npm i -D @deploysentinel/playwright-watch @deploysentinel/playwright
> ```
> 
> Run:
> 
> ```
> npx pw-watch --reporter @deploysentinel/playwright
> ```

Nice for checking, thanks!

**Comment by RIP21** [comment]
Please implement that!
My `global-setup.ts` takes easily 10-20 seconds (starts DB and Redis container, Backend, and Next.js Frontend) but then is fine to be reused as tests are isolated.
Restarting all this machinery over and over again takes a lot of time and having a watch mode that is not starting the `global-setup.ts` would be super nice.

I think I can work around it by starting it all on some unique predefined ports before e2e starts. But that won't be as seamless and super reliable as having always absolutely clean test containers running on random ports. (I use `testcontainers-node` to start it all)

**Comment by aslushnikov** [comment]
This is now fixed with the new UI mode in Playwright 1.32. 
- see https://github.com/microsoft/playwright/releases/tag/v1.32.0 for details
- video demo of the UI mode: https://www.youtube.com/watch?v=jF0yA-JLQW0

Please give UI mode a try and let us know (via filing new issues) if something doesn't work!

**Comment by RichiCoder1** [comment]
That's fantastic but...is there a way to watch without the UI? There's reasons launch a full UI just to watch tests and get feedback might not be idea.

**Comment by aslushnikov** [comment]
@RichiCoder1  would you mind filing a separate issue and explaning your use case there? This way we'll be able to gauge interest and asses uses cases.

Thanks!

**Comment by RichiCoder1** [comment]
Created https://github.com/microsoft/playwright/issues/21960 for headless watch mode. Lmk if I can phrase it better!

**Comment by dman777** [comment]
Just to mention, I do not use VS code (I use VIM) so if this feature is ever implemented please do not tightly couple it with VS code.

**Comment by luchillo17** [comment]
I'm wondering if there's a way to make the playwright e2e watch both the tests and the source code, I know it might not be a priority since the usual SDLC even with TDD is to have unit tests running on watch mode while we develop, and e2e gets added later at the end just to tie it all together, or to cover dark bugs regression testing.

**Comment by saneatsu** [comment]
Try the following command.

ref: https://github.com/microsoft/playwright/issues/21960#issuecomment-1483604692

```sh
$ PWTEST_WATCH=1 npx playwright test
```
